backup-stores/s3/src/main/java/io/camunda/zeebe/backup/s3/S3BackupStore.java

```diff
@@ -156,25 +154,27 @@ public CompletableFuture<Backup> restore(final BackupIdentifier id, final Path t
     LOG.info("Restoring {} to {}", id, targetFolder);
     final var backupPrefix = objectPrefix(id);
     return requireBackupStatus(id, EnumSet.of(BackupStatusCode.COMPLETED))
-        .thenComposeAsync(this::readMetadataObject)
+        .thenComposeAsync(this::readManifestObject)
         .thenComposeAsync(
-            metadata ->
+            manifest ->
                 downloadNamedFileSet(
-                        backupPrefix + SEGMENTS_PREFIX, metadata.segmentFileNames(), targetFolder)
+                        backupPrefix + SEGMENTS_PREFIX, manifest.segmentFileNames(), targetFolder)
                     .thenCombineAsync(
                         downloadNamedFileSet(
                             backupPrefix + SNAPSHOT_PREFIX,
-                            metadata.snapshotFileNames(),
+                            manifest.snapshotFileNames(),
                             targetFolder),
                         (segments, snapshot) ->
-                            new BackupImpl(id, metadata.descriptor(), snapshot, segments)));
+                            new BackupImpl(id, manifest.descriptor(), snapshot, segments)));
   }
 
   @Override
   public CompletableFuture<BackupStatusCode> markFailed(
       final BackupIdentifier id, final String failureReason) {
     LOG.info("Marking {} as failed", id);
-    return setStatus(id, new Status(BackupStatusCode.FAILED, Optional.of(failureReason)));
+    return readManifestObject(id)
+        .thenComposeAsync(manifest -> writeManifestObject(manifest.withFailedStatus(failureReason)))
+        .thenApply(Manifest::statusCode);
```

Comment:
:x: If manifest does not exist, or if there is an error reading it then we cannot mark it as failed. This is not ideal. Backup can be marked as failure even before writing the manifest. For example, when no valid snapshots are found, BackupManager will mark this backup as failed even before writing any objects to the backup store.

That case must be handled here - if there is no manifest, it should create an incomplete manifest with known information and set the status to fail. Then we must verify that we can read from an incomplete manifest without any issues.

Reply:
Thanks for pointing this out, I completely missed this :+1: 

I've included (another) refactoring that allows for type checked manifests, more details in the commit message: 5b0a7730f2f0355febb538f7b32ba7c8ddeb2e22

There, I also fixed this issue and added a test here: cba12bb185f9148a1e075f05ddf679bb5b7ab9a7